### PR TITLE
Added support for importing of http(s):// URLs (resolves #722)

### LINKED ISSUE #722: Add support for importing/exporting of http(s):// URLs

### PR DESCRIPTION
**Pull request by JakeNarkizian**
Support was added by adding functionality to AbstractJobStore._supportsUrl, AbstractJobStore._readFromUrl.

### CODE REVIEW

**Comment by JakeNarkizian** [comment]
Used the the code at [this repo](https://github.com/JakeNarkizian/Real-Time-Benchmark) to run benchmarks. Urllib2 was the clear winner. Here is the output from the benchmark:

function, download size(bytes), time(seconds)
curl_meth, 32768, 0.0216031074524
curl_meth, 32769, 0.0138919353485
curl_meth, 32770, 0.0135269165039
curl_meth, 32771, 0.0146420001984
curl_meth, 32772, 0.0142569541931
curl_meth, 32773, 0.0140700340271
curl_meth, 32774, 0.0131349563599
curl_meth, 32775, 0.0137469768524
curl_meth, 32776, 0.0147151947021
curl_meth, 32777, 0.0138819217682
curl_meth, 32778, 0.0143828392029
curl_meth, 32779, 0.0139961242676
curl_meth, 32780, 0.0139060020447
curl_meth, 32781, 0.0137798786163
curl_meth, 32782, 0.0133731365204
curl_meth, 32783, 0.0133800506592
curl_meth, 32784, 0.0133140087128
curl_meth, 32785, 0.0142619609833
curl_meth, 32786, 0.0132369995117
curl_meth, 32787, 0.0140080451965
curl_meth, 32788, 0.0127730369568
curl_meth, 32789, 0.0141589641571
curl_meth, 32790, 0.0142722129822
curl_meth, 32791, 0.0134489536285
curl_meth, 32792, 0.0137810707092
curl_meth, 32793, 0.0135860443115
curl_meth, 32794, 0.0139169692993
curl_meth, 32795, 0.0138819217682
curl_meth, 32796, 0.0145030021667
curl_meth, 32797, 0.0128970146179
curl_meth, 32798, 0.0137360095978
curl_meth, 32799, 0.013787984848
curl_meth, 32800, 0.0154118537903
curl_meth, 32801, 0.0138709545135
curl_meth, 32802, 0.0138070583344
curl_meth, 32803, 0.0137519836426
curl_meth, 32804, 0.0139529705048
curl_meth, 32805, 0.0145149230957
curl_meth, 32806, 0.0130860805511
curl_meth, 32807, 0.0135459899902
curl_meth, 32808, 0.0134999752045
curl_meth, 32809, 0.0135970115662
curl_meth, 32810, 0.0131549835205
curl_meth, 32811, 0.0140161514282
curl_meth, 32812, 0.0135819911957
curl_meth, 32813, 0.0133218765259
curl_meth, 32814, 0.0137050151825
curl_meth, 32815, 0.013689994812
curl_meth, 32816, 0.0138349533081
curl_meth, 32817, 0.0131959915161
requests_meth, 32768, 0.0114939212799
requests_meth, 32769, 0.00448513031006
requests_meth, 32770, 0.0040910243988
requests_meth, 32771, 0.00403785705566
requests_meth, 32772, 0.0042929649353
requests_meth, 32773, 0.00430083274841
requests_meth, 32774, 0.00399804115295
requests_meth, 32775, 0.00568294525146
requests_meth, 32776, 0.00471901893616
requests_meth, 32777, 0.00431180000305
requests_meth, 32778, 0.00390696525574
requests_meth, 32779, 0.00437688827515
requests_meth, 32780, 0.0046911239624
requests_meth, 32781, 0.00457215309143
requests_meth, 32782, 0.00489592552185
requests_meth, 32783, 0.00427889823914
requests_meth, 32784, 0.00405383110046
requests_meth, 32785, 0.00456309318542
requests_meth, 32786, 0.00461196899414
requests_meth, 32787, 0.00414204597473
requests_meth, 32788, 0.00409293174744
requests_meth, 32789, 0.00512194633484
requests_meth, 32790, 0.00506210327148
requests_meth, 32791, 0.00442099571228
requests_meth, 32792, 0.00437593460083
requests_meth, 32793, 0.00525212287903
requests_meth, 32794, 0.00474810600281
requests_meth, 32795, 0.00408887863159
requests_meth, 32796, 0.00411295890808
requests_meth, 32797, 0.00404000282288
requests_meth, 32798, 0.00409984588623
requests_meth, 32799, 0.00410485267639
requests_meth, 32800, 0.00409388542175
requests_meth, 32801, 0.00424194335938
requests_meth, 32802, 0.00411105155945
requests_meth, 32803, 0.0044538974762
requests_meth, 32804, 0.00433301925659
requests_meth, 32805, 0.00459790229797
requests_meth, 32806, 0.00419783592224
requests_meth, 32807, 0.004802942276
requests_meth, 32808, 0.00395011901855
requests_meth, 32809, 0.00433397293091
requests_meth, 32810, 0.00488781929016
requests_meth, 32811, 0.00408816337585
requests_meth, 32812, 0.00437307357788
requests_meth, 32813, 0.00404500961304
requests_meth, 32814, 0.00436401367188
requests_meth, 32815, 0.00448989868164
requests_meth, 32816, 0.00418496131897
requests_meth, 32817, 0.00430512428284
urllib_meth, 32768, 0.00796604156494
urllib_meth, 32769, 0.00186800956726
urllib_meth, 32770, 0.00188302993774
urllib_meth, 32771, 0.00196385383606
urllib_meth, 32772, 0.00205206871033
urllib_meth, 32773, 0.00260496139526
urllib_meth, 32774, 0.0019268989563
urllib_meth, 32775, 0.00187802314758
urllib_meth, 32776, 0.0022120475769
urllib_meth, 32777, 0.0020489692688
urllib_meth, 32778, 0.00192284584045
urllib_meth, 32779, 0.00183987617493
urllib_meth, 32780, 0.00242900848389
urllib_meth, 32781, 0.00192022323608
urllib_meth, 32782, 0.00276207923889
urllib_meth, 32783, 0.0021059513092
urllib_meth, 32784, 0.00182890892029
urllib_meth, 32785, 0.00183510780334
urllib_meth, 32786, 0.00195813179016
urllib_meth, 32787, 0.00230193138123
urllib_meth, 32788, 0.0019519329071
urllib_meth, 32789, 0.00213980674744
urllib_meth, 32790, 0.00261306762695
urllib_meth, 32791, 0.0026261806488
urllib_meth, 32792, 0.00248599052429
urllib_meth, 32793, 0.00210094451904
urllib_meth, 32794, 0.00225210189819
urllib_meth, 32795, 0.00234484672546
urllib_meth, 32796, 0.00192809104919
urllib_meth, 32797, 0.00185084342957
urllib_meth, 32798, 0.00198698043823
urllib_meth, 32799, 0.00194692611694
urllib_meth, 32800, 0.00189590454102
urllib_meth, 32801, 0.00190997123718
urllib_meth, 32802, 0.00181794166565
urllib_meth, 32803, 0.0019359588623
urllib_meth, 32804, 0.00205302238464
urllib_meth, 32805, 0.00225710868835
urllib_meth, 32806, 0.00203800201416
urllib_meth, 32807, 0.00210094451904
urllib_meth, 32808, 0.00206589698792
urllib_meth, 32809, 0.00188088417053
urllib_meth, 32810, 0.00208711624146
urllib_meth, 32811, 0.00200510025024
urllib_meth, 32812, 0.00213408470154
urllib_meth, 32813, 0.00190401077271
urllib_meth, 32814, 0.00188589096069
urllib_meth, 32815, 0.00197410583496
urllib_meth, 32816, 0.00186681747437
urllib_meth, 32817, 0.00178384780884

**Comment by JakeNarkizian** [comment]
Averages for the above data are as follows:
```
function, time(seconds)
urllib_meth, 0.0068893734614048
curl_meth , 0.013947839736938002
requests_meth, 0.0092427372932429
```

**Comment by JakeNarkizian** [comment]
@hannes-ucsc this is ready for review

File: src/toil/jobStores/abstractJobStore.py [hunk 1]
```diff
@@ -270,19 +275,19 @@ def _exportFile(self, otherCls, jobStoreFileID, url):
         with self.readFileStream(jobStoreFileID) as readable:
             otherCls._writeToUrl(readable, url)
 
-    @abstractclassmethod
+    @classmethod
```
Comment (hannes-ucsc) [inline src/toil/jobStores/abstractJobStore.py]: Let's leave this abstract and push your implementation down into a JobStoreSupport class in the same module as the AbstractJobStore.

```
class JobStoreSupport(AbstractJobStore):
…
```

File: src/toil/jobStores/abstractJobStore.py [hunk 2]
```diff
@@ -292,13 +297,13 @@ def _writeToUrl(cls, readable, url):
...
                 supported URL scheme e.g. a blob in an Azure Blob Storage container.
         :param readable: a readable stream
         """
         raise NotImplementedError()
 
-    @abstractclassmethod
+    @classmethod
     def _supportsUrl(cls, url):
         """
         Returns True if the job store supports the URL's scheme.
```
Comment (hannes-ucsc) [inline src/toil/jobStores/abstractJobStore.py]: `… for importing from or exporting to.`

We also need to discuss how to represent the case where a URL is supported for import only.

File: src/toil/jobStores/abstractJobStore.py [hunk 3]
```diff
@@ -308,7 +313,7 @@ def _supportsUrl(cls, url):
         :param urlparse.ParseResult url: a parsed URL that may be supported
         :return bool: returns true if the cls supports the URL
         """
-        raise NotImplementedError()
+        return url.scheme.lower() == 'http' or url.scheme.lower() == 'https'
```
Comment (hannes-ucsc) [inline src/toil/jobStores/abstractJobStore.py]: Use `in` operator with a tuple on the right side.

File: src/toil/test/jobStores/jobStoreTest.py [hunk 4]
```diff
@@ -377,6 +377,17 @@ def importExportFile(self, otherJobStore, size):
...
 
+        def testImportHttpUrl(self):
+            """
+            Imports an arbitrary http url and checks its hash.
+            """
+            srcUrl = 'https://raw.githubusercontent.com/BD2KGenomics/toil/master/Makefile'
+            srcHash = hashlib.md5(urllib2.urlopen(srcUrl).read()).hexdigest()
+
+            # test import
+            jobStoreFileID = self.master.importFile(srcUrl)
```
Comment (hannes-ucsc) [inline src/toil/test/jobStores/jobStoreTest.py]: Shouldn't this be tested generatively like the other job store classes?

**Comment by JakeNarkizian** [comment]
Here is raw data from benchmarks relevant to this PR (The code for which can be found [here](https://github.com/JakeNarkizian/Python-Real-Time-Benchmark-Tool/blob/master/main.py))

```
function,requests,curlBuffered,curlUnbuffered,urllib
2^0,0.0102639198303,0.0154399871826,0.0134329795837,0.00829315185547
2^5,0.00409388542175,0.013396024704,0.013787984848,0.00189709663391
2^10,0.0040910243988,0.0129778385162,0.0137050151825,0.00185298919678
2^15,0.00421810150146,0.0141730308533,0.0143661499023,0.00196409225464
2^20,0.00729703903198,0.016853094101,0.0175499916077,0.00703191757202
2^25,0.0892179012299,0.0745449066162,0.0780229568481,0.0845441818237
2^30,2.39939713478,6.695291996,2.08755493164,2.5330529213
```

Here is a graph of the data (lower is better)

Alt Text

urllib2 looks like the winner

File: src/toil/jobStores/abstractJobStore.py [hunk 5]
```diff
@@ -108,18 +111,20 @@ def getJobStoreClasses():
             jobStoreClasses.append(getattr(module, className))
     return jobStoreClasses
 
+methods = Enum('importOnly', 'exportOnly', 'both')
```
Comment (hannes-ucsc) [inline src/toil/jobStores/abstractJobStore.py]: s/method/mode because method is overloaded (it means something for OO and HTTP)

Comment (hannes-ucsc) [inline src/toil/jobStores/abstractJobStore.py]: s/methods/mode

(Don't use plural)

File: src/toil/jobStores/abstractJobStore.py [hunk 6]
```diff
@@ -866,3 +884,28 @@ def _defaultTryCount(self):
     @classmethod
     def _validateSharedFileName(cls, sharedFileName):
         return bool(cls.sharedFileNameRegex.match(sharedFileName))
+
+
+class SupportJobStore(AbstractJobStore):
```
Comment (hannes-ucsc) [inline src/toil/jobStores/abstractJobStore.py]: JobStoreSupport

File: src/toil/jobStores/abstractJobStore.py [hunk 7]
```diff
@@ -866,3 +884,28 @@ def _defaultTryCount(self):
...
+        def importOnly():
+            return url.scheme.lower() in ('http', 'https')
+
+        def exportOnly():
+            return False
+
+        def both():
+            return False
+
+        return cls._supportsUrlBody(method, importOnly, exportOnly, both)
```
Comment (hannes-ucsc) [inline src/toil/jobStores/abstractJobStore.py]: Much to complicated. _supportsUrlBody needs to go, sorry. Maybe Enum wasn't the right decision.

Do we really ever need to know if a URL is supported for both import and export? We are either importing or exporting, so when do we care if a batch system supports both modes?

File: src/toil/test/__init__.py [hunk 8]
```diff
@@ -394,23 +394,31 @@ def insertMethodToClass():
...
-        left[pvName] = {prmName: val}
-
-    # get cartesian product
-    while len(kwargs) > 0:
-        permuteIntoLeft(left, *pop(kwargs))
-
-    # set class attributes
-    targetClass = targetClass or generalMethod.im_class
-    for prmNames, prms in left.items():
+    if len(kwargs) > 0:
```
Comment (hannes-ucsc) [inline src/toil/test/__init__.py]: This is an unrelated change, correct? If so, separate commit please.

File: src/toil/test/jobStores/jobStoreTest.py [hunk 9]
```diff
@@ -846,3 +861,4 @@ class EncryptedAzureJobStoreTest(AzureJobStoreTest, hidden.AbstractEncryptedJobS
     pass
 
 hidden.AbstractJobStoreTest.makeImportExportTests()
+hidden.AbstractJobStoreTest.makeImportOnlyTests()
```
Comment (hannes-ucsc) [inline src/toil/test/jobStores/jobStoreTest.py]: NL missing at EOL

File: setup.py [hunk 10]
```diff
@@ -24,7 +24,8 @@
     author_email='benedict@soe.usc.edu',
     url="https://github.com/BD2KGenomics/toil",
     install_requires=[
-        'bd2k-python-lib==1.13.dev14'],
+        'bd2k-python-lib==1.13.dev14',
+        'enum=0.4.6'],
```
Comment (hannes-ucsc) [inline setup.py]: ==

File: src/toil/test/__init__.py (hunk 8 above)
Comment (JakeNarkizian) [inline src/toil/test/__init__.py]: It is related to the test generation of the http tests

File: src/toil/jobStores/abstractJobStore.py [hunk 11]
```diff
@@ -866,3 +884,28 @@ def _defaultTryCount(self):
...
+        def importOnly():
+            return url.scheme.lower() in ('http', 'https')
+
+        def exportOnly():
+            return False
+
+        def both():
+            return False
+
+        return cls._supportsUrlBody(mode, importOnly, exportOnly, both)
```
Comment (hannes-ucsc) [inline src/toil/jobStores/abstractJobStore.py]: Much to complicated. _supportsUrlBody needs to go, sorry. Maybe Enum wasn't the right decision.

Do we really ever need to know if a URL is supported for both import and export? We are either importing or exporting, so when do we care if a batch system supports both modes?

File: src/toil/test/__init__.py (hunk 8 above)
Comment (hannes-ucsc) [inline src/toil/test/__init__.py]: Why did the test generation have to change?

Comment (JakeNarkizian) [inline src/toil/test/__init__.py]: The import only tests (which currently only tests for http urls) does not require any parameter permutations because the inheriting jobStore class will just import from some http url. The change in make_tests() allows a test method to be generated without any general method parametrization  
```
 @classmethod
        def makeImportOnlyTests(cls):
            def importHttpFile(self):
                # prepare random file for import
                self.master.partSize = cls.mpTestPartSize
                srcUrl = 'https://raw.githubusercontent.com/BD2KGenomics/toil/master/Makefile'
                srcHash = hashlib.md5(urllib2.urlopen(srcUrl).read()).hexdigest()
                # test import
                jobStoreFileID = self.master.importFile(srcUrl)
                self.assertEqual(self._hashJobStoreFileID(jobStoreFileID), srcHash)

            make_tests(importHttpFile, targetClass=cls)
```

Comment (hannes-ucsc) [inline src/toil/test/__init__.py]: Makes sense.

File: src/toil/jobStores/abstractJobStore.py (hunk 11 above)
Comment (JakeNarkizian) [inline src/toil/jobStores/abstractJobStore.py]: I agree, this is a lot of code for what looks to me like more of an edge case. If these changes are reversed and say this happens
```
with Toil(options) as toil:
    toil.exportFile(someJobStoreFileID, 'http://someURL"')
```
A not implemented error will be raised because jobStoreSupport() lacks ``` _writeToUrl()```. Since this exception will be the one ultimately seen by the user, what do you think of raising a more specific exception than not implemented, something along the lines of ``` UnsupportedExportError```, assuming also _readFromUrl would raise ``` UnsuportedImportError```

Comment (hannes-ucsc) [inline src/toil/jobStores/abstractJobStore.py]: The complexity you introduced stems from the fact that `both` implies `importOnly` and `exportOnly` (which, BTW, is why I didn't like the `Only` suffix). There are ways around this (e.g. using bitwise operators like in C), but again, we don't need `both`. 

`exportFile(…,'http://…')` should call `_supportsUrl('http://…',export=True)` on each job store class. None of them will return True and a specific exception should be raised, not `NotImplementedError`.

**Comment by JakeNarkizian** [comment]
@hannes-ucsc ready

**Comment by hannes-ucsc** [comment]
So close, but commit message lacks issue reference. I also think the body of the commit message doesn't apply anymore?

**Comment by JakeNarkizian** [comment]
@hannes-ucsc

File: setup.py [hunk 12]
```diff
@@ -24,7 +24,8 @@
     author_email='benedict@soe.usc.edu',
     url="https://github.com/BD2KGenomics/toil",
     install_requires=[
-        'bd2k-python-lib==1.13.dev14'],
+        'bd2k-python-lib==1.13.dev14',
+        'enum==0.4.6'],
```
Comment (hannes-ucsc) [inline setup.py]: What do we need `enum` for?

File: src/toil/jobStores/abstractJobStore.py [hunk 13]
```diff
@@ -250,6 +253,9 @@ def importFile(self, srcUrl):
             - 'file' for local files
                 e.g. file://local/file/path
```
Comment (hannes-ucsc) [inline src/toil/jobStores/abstractJobStore.py]: `file:///` (three slashes, not two)

**Comment by JakeNarkizian** [comment]
Jenkins, test this please.
